### ISSUE TITLE
Inconsistency networks.layer.AffineTransform and transforms.Affine

### ISSUE THREAD

**Issue by Constantin-Jehn** [Project-MONAI/MONAI#4281]
Hello,

**Describe the bug**
Applying the same affine transformation using networks.layer.AffineTransform and transforms.Affine seems to lead to different result. The layer seems to introduce an unwanted shear.

**To Reproduce**
1. Download [10_3T_nody_001.nii.gz.zip](https://github.com/Project-MONAI/MONAI/files/8700120/10_3T_nody_001.nii.gz.zip)
2. Run the following notebook

```python
import monai
from monai.transforms import (
    AddChanneld,
    LoadImaged,
    ToTensord,
)
import torch as t
from copy import deepcopy
```

Load images and copy for comparison

```python
path = "10_3T_nody_001.nii.gz"
add_channel = AddChanneld(keys=["image"])
loader = LoadImaged(keys=["image"])
to_tensor = ToTensord(keys=["image"])
monai_dict = {"image": path}
monai_image_layer = loader(monai_dict)
monai_image_layer = add_channel(monai_image_layer)
monai_image_layer = to_tensor(monai_image_layer)
monai_image_transform = deepcopy(monai_image_layer)
```

Set up example affine

```python
translations = [0,0,0]
rotations = [t.pi/8, 0 ,0]

rotation_tensor = monai.transforms.utils.create_rotate(3, rotations,backend='torch')
translation_tensor = monai.transforms.utils.create_translate(3, translations,backend='torch')
affine =  (t.matmul(translation_tensor,rotation_tensor))
```

Define affine layer and transform

```python
monai_aff_layer = monai.networks.layers.AffineTransform(mode = "bilinear",  normalized = True, align_corners= True, padding_mode = "zeros")
monai_aff_trans = monai.transforms.Affine(affine=affine, norm_coords = True, image_only = True)
```

Do transform using affine layer, add another dimension for batch

```python
input_batched = monai_image_layer["image"].unsqueeze(0)

monai_image_layer["image"] = monai_aff_layer(input_batched ,t.tensor(affine).unsqueeze(0))

monai_image_layer["image"] = monai_image_layer["image"].squeeze().unsqueeze(0)

nifti_saver = monai.data.NiftiSaver(output_dir="tests", output_postfix = "layer",
                                            resample=False, padding_mode="zeros",
                                            separate_folder=False)
nifti_saver.save(monai_image_layer["image"],monai_image_layer["image_meta_dict"])
```

    file written: tests/10_3T_nody_001_layer.nii.gz.

    /var/folders/ll/btjq1lpn2msbmbn34rst5wz80000gn/T/ipykernel_64711/3194165265.py:3: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
      monai_image_layer["image"] = monai_aff_layer(input_batched ,t.tensor(affine).unsqueeze(0))

Do transform using transform.Affine

```python
monai_image_transform["image"] = monai_aff_trans(monai_image_transform["image"], padding_mode="zeros")

nifti_saver = monai.data.NiftiSaver(output_dir="tests", output_postfix = "trans",
                                            resample=False, padding_mode="zeros",
                                            separate_folder=False)
nifti_saver.save(monai_image_transform["image"], monai_image_transform["image_meta_dict"])
```

    file written: tests/10_3T_nody_001_trans.nii.gz.

```python

```

**Expected behavior**
Unambigious behaviour of the network and transforms affine transform.

**Screenshots**
original file:
![original_file](https://user-images.githubusercontent.com/66247479/168592276-191914b6-5848-4c27-aaee-794a70689cd5.png)
using networks.layer:
![networks_layer_Affine](https://user-images.githubusercontent.com/66247479/168593634-efffed38-a531-40e7-9f2f-7fe0e6cd209e.png)
using transforms.Affine
![transforms_Affine](https://user-images.githubusercontent.com/66247479/168592301-13c943ec-aa36-4fe7-8baa-612a5b8d1da4.png)

**Environment**

================================
Printing MONAI config...
================================
MONAI version: 0.8.1+181.ga676e387
Numpy version: 1.20.3
Pytorch version: 1.10.2
MONAI flags: HAS_EXT = False, USE_COMPILED = False
MONAI rev id: a676e3876903e5799eb3d24245872ea4f3a5b152
MONAI __file__: /Users/constantin/opt/anaconda3/envs/SVR_pytorch/lib/python3.9/site-packages/monai/__init__.py

Optional dependencies:
Pytorch Ignite version: NOT INSTALLED or UNKNOWN VERSION.
Nibabel version: 3.2.2
scikit-image version: 0.18.3
Pillow version: 8.4.0
Tensorboard version: 2.9.0
gdown version: NOT INSTALLED or UNKNOWN VERSION.
TorchVision version: 0.11.3
tqdm version: 4.62.3
lmdb version: NOT INSTALLED or UNKNOWN VERSION.
psutil version: 5.8.0
pandas version: 1.3.4
einops version: NOT INSTALLED or UNKNOWN VERSION.
transformers version: NOT INSTALLED or UNKNOWN VERSION.
mlflow version: NOT INSTALLED or UNKNOWN VERSION.

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies

================================
Printing system config...
================================
System: Darwin
Mac version: 10.16
Platform: macOS-10.16-x86_64-i386-64bit
Processor: i386
Machine: x86_64
Python version: 3.9.7
Process name: python3.9
Command: ['python', '-c', 'import monai; monai.config.print_debug_info()']
Open files: []
Num physical CPUs: 8
Num logical CPUs: 8
Num usable CPUs: UNKNOWN for given OS
CPU usage (%): [32.3, 32.3, 27.3, 25.2, 10.3, 57.9, 2.4, 1.6]
CPU freq. (MHz): 2400
Load avg. in last 1, 5, 15 mins (%): [23.4, 24.4, 22.5]
Disk usage (%): 54.7
Avg. sensor temp. (Celsius): UNKNOWN for given OS
Total physical memory (GB): 16.0
Available memory (GB): 0.7
Used memory (GB): 1.2

================================
Printing GPU config...
================================
Num GPUs: 0
Has CUDA: False
cuDNN enabled: False

**Comment by wyli** [comment]
This is a good question, in your example the affine of `monai.networks.layers.AffineTransform` is interpreted as applying it to the normalized coordinates (in the range of [-1, 1]) while `monai.transforms.Affine`'s affine is interpreted as to the `create_grid` coordinates ([-d/2, d/2] where d is the length of the dimension) https://github.com/Project-MONAI/MONAI/blob/f4ff8e2a07b5af29a62ef6ed14b92bd6806adc61/monai/transforms/utils.py#L630-L632

So, to get consistent outputs, you can modify the `affine` of `monai.transforms.Affine` with a scaling transform (assuming input image shape [99, 92, 69] as the one you provided):
```py
s = t.diag(t.as_tensor([2 / (99 - 1.0), 2 / (92 - 1.0), 2 / (69 - 1.0), 1.0]))                                          
new_affine = affine @ s                                                                                                 
monai_aff_trans = monai.transforms.Affine(affine=new_affine, norm_coords = False, image_only = True)
```

**Comment by Constantin-Jehn** [comment]
Thanks a lot for your reply.
Indeed, your answers makes the two transforms consistent.

However, this approach seems to introduce a shear to the image. And changing the 'norm_coords' parameter to False in the layer leads to a rotation which is not centered at the image center like the following (still from same file).

![layer_norm_coords_false](https://user-images.githubusercontent.com/66247479/168756687-522a925e-1e8c-411b-97c1-e5765092befd.png)

To give some context, I am working on compensation of motion corruption where I want to optimize transform parameters. Therefore, I need to use the layer as it allows backpropagation.

Comparing to the affine transform from torchio for example which is based on sitk afaik, the result of transforms.Affine is much closer (but not consistent though): 
From the notebook:
```python
import monai
from monai.transforms import (
    AddChanneld,
    LoadImaged,
    ToTensord,
)
import torch as t
from copy import deepcopy
import torchio as tio
import numpy as np
```


```python
rotations = [t.pi/8, 0 ,0]

tio_image = tio.Image("cropped_images/10_3T_nody_001.nii.gz")

rot_deg = np.array(rotations) * 180 / np.pi
tio_rotations = (rot_deg[0], rot_deg[1], rot_deg[2])

tio_aff = tio.Affine(1,tio_rotations,0,center="image",default_pad_value=0.0, image_interpolation="linear")
```
![torchio_transform](https://user-images.githubusercontent.com/66247479/168757997-56460155-bfc1-4b9f-93ec-39d92c68fed1.png)

Ideally, someone could explain me how to get consistency with the transform from torchio as well, as it offers better interpolation modes, which I could use after optimization for updating the volume.

Cheers.

**Comment by Constantin-Jehn** [comment]
Thanks again for you comment.
As the affine is optimized for each image individually, I would have thought there is no harm if it would be coupled to the spatial size.

Any thoughts on consistency with torchio/sitk affine transform?

As a general remark, I think it would be benefitial to make options in layers.AffineTransform more constitent with transform.Affine.

**Comment by wyli** [comment]
> consistency with torchio/sitk affine transform?

I'm not familiar with torchio. if you could make a working example I can look into the gaps.

if the gaps are small, one possibility is the precision of `affine`, you can try `torch.float64` for the `affine` with monai's api.

**Comment by Constantin-Jehn** [comment]
I edited my comment "Thanks a lot for your reply.
Indeed, your answers makes the two transforms consistent....."

such that the notebook actually also applies the transform and saves the output.
As posted there as Screenshot the difference is rather significant.

Thanks, for any further support.

**Comment by wyli** [comment]
thanks, I can run the code, the difference is mainly in the initial affine information. in the following example the output would be very similar (if the tio_image is initialized with an identity affine):

```py
from copy import deepcopy

import numpy as np
import torch as t
import torchio as tio

from monai.transforms import AddChanneld, LoadImaged, ToTensord

path = "10_3T_nody_001.nii.gz"
add_channel = AddChanneld(keys=["image"])
loader = LoadImaged(keys=["image"])
to_tensor = ToTensord(keys=["image"])
monai_dict = {"image": path}
monai_image_layer = loader(monai_dict)
monai_image_layer = add_channel(monai_image_layer)
monai_image_layer = to_tensor(monai_image_layer)
monai_image_transform = deepcopy(monai_image_layer)
tio_image = deepcopy(monai_image_layer)


rotations = [0, 0, t.pi / 8]
# changed here:
tio_image = tio.Image(tensor=tio_image['image'][0][..., None])

rot_deg = np.array(rotations) * 180 / np.pi
tio_rotations = (rot_deg[0], rot_deg[1], rot_deg[2])
tio_aff = tio.Affine(1, tio_rotations, 0, center="image", default_pad_value=0.0, image_interpolation="linear")
tio_trans = tio_aff(tio_image)
tio_trans.save("tests/tio_lin_resample.nii.gz")
```

**Comment by Constantin-Jehn** [comment]
Hi,

Unfortunately, I cannot replicate getting similar outputs as suggested (Maybe you could provide Screenshots?).

Also I don't understand, how one could/should just change the affine matrix. Essentially, the image is not the same anymore with different affine (it's somewhere else in space, with other spacing etc.).

I am curious to understand your approach.

Cheers.

**Comment by wyli** [comment]
sure, I attach the full script based on your example here: 
- it requires my pull request branch to run https://github.com/wyli/MONAI/tree/4281-norm-mode 
- sitk and monai's parameter conventions are different, for the first two axes there is `t.pi/8` vs `-t.pi/8` in the script.
- because we drop the initial affine in the script, so the saved testing outcome couldn't be properly visualized along with the original image input.

```py
from copy import deepcopy

import numpy as np
import torch as t
import torchio as tio

import monai
from monai.transforms import AddChanneld, LoadImaged, ToTensord

path = "10_3T_nody_001.nii.gz"
add_channel = AddChanneld(keys=["image"])
loader = LoadImaged(keys=["image"])
to_tensor = ToTensord(keys=["image"])
monai_dict = {"image": path}
monai_image_layer = loader(monai_dict)
monai_image_layer = add_channel(monai_image_layer)
monai_image_layer = to_tensor(monai_image_layer)
monai_image_transform = deepcopy(monai_image_layer)
tio_image = deepcopy(monai_image_layer)


translations = [0, 0, 0]
rotations = [0, t.pi / 8, 0]

rotation_tensor = monai.transforms.utils.create_rotate(3, rotations, backend="torch")
translation_tensor = monai.transforms.utils.create_translate(3, translations, backend="torch")
affine = t.matmul(translation_tensor, rotation_tensor)


monai_aff_layer = monai.networks.layers.AffineTransform(
    mode="bilinear", normalized=False, align_corners=True, padding_mode="zeros", zero_centered=True
)
monai_aff_trans = monai.transforms.Affine(affine=affine, normalized=False, image_only=True)


input_batched = monai_image_layer["image"].unsqueeze(0)
monai_image_layer["image"] = monai_aff_layer(input_batched, t.tensor(affine).unsqueeze(0))
monai_image_layer["image"] = monai_image_layer["image"].squeeze().unsqueeze(0)
nifti_saver = monai.data.NiftiSaver(
    output_dir="tests", output_postfix="layer", resample=False, padding_mode="zeros", separate_folder=False
)
monai_image_layer["image_meta_dict"].pop("affine")
nifti_saver.save(monai_image_layer["image"], monai_image_layer["image_meta_dict"])


monai_image_transform["image"] = monai_aff_trans(monai_image_transform["image"], padding_mode="zeros")
nifti_saver = monai.data.NiftiSaver(
    output_dir="tests", output_postfix="trans", resample=False, padding_mode="zeros", separate_folder=False
)
monai_image_transform["image_meta_dict"].pop("affine")
nifti_saver.save(monai_image_transform["image"], monai_image_transform["image_meta_dict"])


rotations = [0, -t.pi / 8, 0]
tio_image = tio.Image(tensor=tio_image["image"])
rot_deg = np.array(rotations) * 180 / np.pi
tio_rotations = (rot_deg[0], rot_deg[1], rot_deg[2])
tio_aff = tio.Affine(1, tio_rotations, 0, center="image", default_pad_value=0.0, image_interpolation="linear")
tio_trans = tio_aff(tio_image)
tio_trans.save("tests/tio_lin_resample.nii.gz")

```

**Comment by wyli** [comment]
>  how one could/should just change the affine matrix. Essentially, the image is not the same anymore with different affine (it's somewhere else in space, with other spacing etc.).

we are developing a core data structure (MetaTensor) to have the metadata fields
https://github.com/Project-MONAI/MONAI/blob/f4ff8e2a07b5af29a62ef6ed14b92bd6806adc61/monai/data/meta_tensor.py#L45-L55

eventually all the monai components will be aware of the metadata (including the affine), and the metadata will be updated accordingly/automatically within the image processing workflows.

**Comment by Constantin-Jehn** [comment]
This will be a major step ahead - I'm looking forward to that.
For me that's for example a reason to use mainly torchio for preprocessing.

**Comment by wyli** [comment]
> This will be a major step ahead - I'm looking forward to that. For me that's for example a reason to use mainly torchio for preprocessing - as mentioned somewhere else the SpatialCrop transform is very problematic in that regard.

great feedback, thanks @Constantin-Jehn!
